### ISSUE TITLE
fanatec csl dd + csl pedal support

### ISSUE THREAD

**Issue by ciscoyo** [gotzl/hid-fanatecff#42]
hi,
i've got the ready to race bundle csl dd + p1 v2 wheel + normal pedals.
i installed oversteer for testing and ffb seems to work, steering recognises and all buttons work. Great work so far!
(i wanted to test it ingame but raceroom doesen't even start and f1 23 crashes after i power on the wheel).

but the pedals are not working correct. noticed when pushed, late and only sometimes, but that's it. there is no constant screening of it.

i'm offering my help with it ofc. even a discord-call. i'm sure i can track the usb traffic with usbmon or something and share it. it's your call.
cheers

**Comment by gotzl** [comment]
Hi,

could you check with `evtest`? Does a pedal push register also only sometimes/late?
Does this command `/usr/bin/evdev-joystick --evdev $devnode --deadzone 0 --fuzz 0` change anything?

About your comment about F1 23: I think you should turn on the wheel before starting the game. In principle, the game should work. At least I could successfully play some days ago....

**Comment by ZakMcKrack3n** [comment]
@ciscoyo I got the same bundle , for me codemasters Dirt Rally 2.0 started crashing some time ago when in CSW 2.5 compatibility mode (orange led), but works fine in native mode (red LED on base power button).

Dirt Rally 2.0 also crashes when turning on the device mid game or switching into CSW 2.5 compatibility mode , where the native mode works fine, maybe you can test the different modes ? (switching is done by a short press of the power button after turning it on and its finished calibrating itself).

As I have the same pedals, are the pedals working in other games / windows ?
If you have no windows to test and this is your only set up: 
Are the pedals connected to the correct port , otherwise funky things could occur if plugged in to hand brake port etc.

**Comment by dtran11** [comment]
I am seeing the same problem with the pedals in Oversteer with a GT DD Pro. However I think it's just an issue with Oversteer because I see correct values with evtest.

By the way, is the GT DD Pro supported?  I assume so because a lot is working but just didn't see it explicitly in the README. Maybe the internals are the same as a CSL DD?

0eb7:0020 Endor AG FANATEC Wheel

Thanks

**Comment by gotzl** [comment]
@dtran11 looks like the 'Pro' is just the torque update? As you show, the PID does not change from the non-Pro one. So, I would say it is 'supported' and I'll update the readme. (I put 'supported' in quotes because only the wheelbase is supported by this driver (FFB), while the wheel itself with its display and some of its buttons is not.)

**Comment by ZakMcKrack3n** [comment]
@gotzl the "Pro" actually has the chip in the base to enable playstation support, it also has both 5/8Nm power supply options.

Probably works identical to a CSL DD pro (I got the identical usb ids) , maybe we can update the readme that this is no longer experimental, no complains here with the CSL DD at least.

**Comment by ZakMcKrack3n** [comment]
> raceroom dosen't have proton support anymore sadly, so it's not related to you gotzl/hid-fanatecff.

@ciscoyo  Actually you can force the use of proton for any game (under compatibility in the games list) .
I tested raceroom using proton experimental only some weeks ago, ffb worked with csw 2.5 (yellow) mode , did not test with native mode at the time.
As I nearly ran out of space because all of the racing games that work, its currently backed up on an external HDD :)

**Comment by gotzl** [comment]
Hmm.. Missing `/sys/module/hid_fanatec/` indicates that the driver is not properly installed. Note that this directory only exists when the driver got loaded. And the driver loads after plugging-in / powering on the device. To see what's happening when you plugging-in/powering the device look at `journalctl -xfk`. There, you should also see some messages from `hid_fanatec`. You can then also post the output here.

Not sure if I can help much with the games. proton logs are mostly giberish for me as well, unless I know what I'm looking for ^^ However, I'll try to install Raceroom and report back if I also run into problems. And when F1 23 is on sale I'll buy it to check this on my side. What sometimes helps with input-related issues is to disable steam-input ('right-click' -> Properties -> Controller'). Maybe you can try that?

**Comment by gotzl** [comment]
> shit s not looking good..

No, this actually looks good! The `Fanatec FANATEC Wheel: Invalid code 768 type 1` appear AFAIU because the wheel has too many buttons ... It doesn't do any harm. The `ftec_csl_elite 0003:0EB7:0020.0011: Hires timer: period = 2 ms` says that everything is fine and the FFB should function.

What you see when you start F1 looks troublesome, though.. There a bunch of IO errors... I'm wondering, what is `/dev/sr0`? Usually `srX` is for optical devices.. These IO errors appera only when the wheel is plugged-id/powered-on?

**Comment by ZakMcKrack3n** [comment]
@gotzl regarding the time period:
I experimented with 1ms (1000Hz , yay) and some games worked (Automobilista 2 for example) fine , but assetto corsa did crash everytime when force feedback was initialised at pre race screens.

Could it be that some kernels have a problem with 2ms as well ?
Input reporting at 1000Hz, but what way ffb events goes through with proton is still a mistery to me.

An idea would be to introduce a module option to be able to modify this hardcoded value for trubleshooting.

The sr0 errors are seem to come from games with active 3rd party copy protection left overs, as stated in the log , there is no medium present , so anything later should not be relevant regarding read errors.

**Comment by gotzl** [comment]
@ZakMcKrack3n oh, that's interesting. I can reproduce the crash of AC. This is very puzzling to me. Technically, the rate of the driver should have no influence on the game what-so-ever...  I will investigate. You're right, this could be related to the problems @ciscoyo has...

**Comment by gotzl** [comment]
Hi, so, F1 23 was on sale and I bought it ;)
For me, the game starts if I use 'default' settings for Steam Input, but it does not recognize the wheel. If I use `Disable Steam Input`, the game doesn't start at all. So, I guess this confirms what @ciscoyo is seeing. I'll create another ticket to track this issue.

**Comment by gotzl** [comment]
So I gave RRE a shot today. After some fiddling, it works for me with FFB. Note that there is a invert FFB button that I had to deselect, but after that, FFB feels alright.

As always, I disabled steam input for the game. Then I had to select the 32bit version, 64bit version wouldn't start because of the shader compilation issue. Then the game loaded but the wheel did not show up. So, I tried the proton63 trick. There, the game loaded and the wheel did show up as well. Back to, proton8, but the wheel did not show up. So, I tried proton7, where the game froze on startup. Then I went to proton-experimental, and TADA!, the game loads, the wheel shows up and FFB works.

So, again a game where the detection of the wheel is problematic for me with proton8, but after switching through some proton versions (w/o removing the pfx) it did eventually work... @ciscoyo so, what I can say for now is that RRE in principle does work. But I don't know the exact recipe how to get the wheel recognized...

**Comment by ZakMcKrack3n** [comment]
> speaking about optical devices, maybe its the steeringwheel leds.

Hard no, all the led endpoints are provided by this project itself, there would not be any without it.
Also under windows, all games use the fanatec middle ware to send telemetry so it can drive the display/leds of the wheel.
That is basically the goal of https://github.com/gotzl/hid-fanatecff-tools , to be a telemetry endpoint and send it to this driver to display gears/speed and rpm led(s).

I suggest if everything is working force feedback and input wise (depending on the game) to close this issue ?

**Comment by melroy89** [comment]
> @ciscoyo I got the same bundle , for me codemasters Dirt Rally 2.0 started crashing some time ago when in CSW 2.5 compatibility mode (orange led), but works fine in native mode (red LED on base power button).
> 
> Dirt Rally 2.0 also crashes when turning on the device mid game or switching into CSW 2.5 compatibility mode , where the native mode works fine, maybe you can test the different modes ? (switching is done by a short press of the power button after turning it on and its finished calibrating itself).
> 
> As I have the same pedals, are the pedals working in other games / windows ? If you have no windows to test and this is your only set up: Are the pedals connected to the correct port , otherwise funky things could occur if plugged in to hand brake port etc.

I have the opposite experience actually.. I have maybe another bundle. But I have [CSL DD base](https://fanatec.com/eu-en/racing-wheels-direct-drive-bases/direct-drive-bases/csl-dd-5-nm) with [CSL Steering Wheel BMW](https://fanatec.com/eu-en/steering-wheels/csl-steering-wheel-bmw) + CSL Pedals (no load cell).

I'm using a handy script using the device ID as input:

```sh
find_by_id(){     v=${1%:*}; p=${1#*:}  # split vid:pid into 2 vars
  v=${v#${v%%[!0]*}}; p=${p#${p%%[!0]*}}  # strip leading zeros
  grep -il "^PRODUCT=$v/$p" /sys/bus/usb/devices/*:*/uevent |     sed s,uevent,, |     xargs -r grep -r '^DEVNAME=' --include uevent; 
}
```

In PC Mode (Red Light), I get: `Bus 007 Device 003: ID 0eb7:0020 Endor AG FANATEC Wheel`, so I use. 

```
find_by_id 0eb7:0020
```

Points me to: `/sys/bus/usb/devices/7-1:1.0/0003:0EB7:0020.000C/`

Driver is:

```
lrwxrwxrwx  1 root root     0 Dec 29 22:37 driver -> ../../../../../../../bus/hid/drivers/ftec_csl_elite/
```

However, games will **not** recognize the wheel nor pedals correctly with the game DiRT4 using Proton/Wine for example.

![image](https://github.com/gotzl/hid-fanatecff/assets/628926/fe90cee4-3902-4286-ba15-e2195f43232b)

HOWEVER, when I set the CSL Base in comparability mode (yellowish/orange light). I will get: `Bus 007 Device 010: ID 0eb7:0004 Endor AG FANATEC Wheel`.

```
find_by_id 0eb7:0004
```

Points me to: `/sys/bus/usb/devices/7-1:1.0/0003:0EB7:0004.000B/`

Driver is also:

```
lrwxrwxrwx  1 root root     0 Dec 29 22:33 driver -> ../../../../../../../bus/hid/drivers/ftec_csl_elite
```

But then the game will recognize my wheel + pedals! And I can play the game without any issues using this driver, but again only in comparability mode (yellow LED light).

![image](https://github.com/gotzl/hid-fanatecff/assets/628926/1256c00c-989d-43e1-b17f-423d6679168b)

Strange right?

**Comment by gotzl** [comment]
@ciscoyo Don't think github offers PMs.. Anyways, thanks for the offer, that's very nice :) But you really don't have to! If you insist though, I use the same nickname for my steam account, you should be able to find it (the one with the serious-sam icon).

**Comment by melroy89** [comment]
> @ciscoyo Don't think github offers PMs.. Anyways, thanks for the offer, that's very nice :) But you really don't have to! If you insist though, I use the same nickname for my steam account, you should be able to find it (the one with the serious-sam icon).

There are multiple steam account with that nickname when searching on `gotzl`, and the ID is not `gotzl`: https://steamcommunity.com/id/gotzl

![image](https://github.com/gotzl/hid-fanatecff/assets/628926/ab53acfd-1476-4ad6-8040-9b7bc5247fc1)

**Comment by dinuxlt** [comment]
Some time ago @gotzl created discord server and it can be used for "fast communication".
Note sure if he is using discord actively.

Or maybe is good idea to add invitation link to README.md ?

**Comment by gotzl** [comment]
> Or maybe is good idea to add invitation link to README.md ?

I see the need for another communication channel as well. I'll check if I can publish the invite to the discord (or create one).

**Comment by gotzl** [comment]
Closing as the CSL DD (Pro) should work.
